ENH: UUID Backward Compatibility Mapping

nyoungbq opened the pull request:
API and UUID updates in order to prep NX for backwards pipeline compatibility

<!-- The text within this markup is a comment, and is intended to provide
guidelines to open a Pull Request for the complex repository. This text will not
be part of the Pull Request. -->


<!-- See the CONTRIBUTING (CONTRIBUTING.md) guide. Specifically:

Start complex commit messages with a standard prefix (and a space):

 * BUG: fix for runtime crash or incorrect result
 * COMP: compiler error or warning fix
 * DOC: documentation change
 * ENH: new functionality
 * PERF: performance improvement
 * STYLE: no logic impact (indentation, comments)
 * WIP: Work In Progress not ready for merge

Provide a short, meaningful message that describes the change you made.

When the PR is based on a single commit, the commit message is usually left as
the PR message.

A reference to a related issue or pull request (https://help.github.com/articles/basic-writing-and-formatting-syntax/#referencing-issues-and-pull-requests)
in your repository. You can automatically
close a related issues using keywords (https://help.github.com/articles/closing-issues-using-keywords/)

@mentions (https://help.github.com/articles/basic-writing-and-formatting-syntax/#mentioning-people-and-teams)
of the person or team responsible for reviewing proposed changes. -->


## Naming Conventions

Naming of variables should descriptive where needed. Loop Control Variables can use `i` if warranted. Most of these conventions are enforced through the clang-tidy and clang-format configuration files.

- [x] Class and Structs are `UpperCamelCase`
- [x] Class private member variables are `m_UpperCamelCase`
- [x] Class methods are `lowerCamelCase`
- [x] Method arguments are `lowerCamelCase`
- [x] Normal variables are `lowerCamelCase`
- [x] Constants are `k_UpperCamelCase`
- [x] Global statics are `s_UpperCamelCase`
- [x] Free Functions are `UpperCamelCase`
- [x] Macros are `ALL_UPPER_SNAKE_CASE`
- [x] Unit test will test data output from the filter.
- [x] Reused strings should be constants in an anonymous namespace
- [x] If parallelization is used, proper use of the abstracted complex classes are used. Using TBB specifically in a filter should be frowned upon unless for a really good reason.

## Filter Checklist

- [x] Parameters should be generally broken down into "Input Parameters", "Required Data Objects", "Created Data Objects". There can be exceptions to this.
- [x] ChoicesParameter selections should be an enumeration defined in the filer header
- [x] Documentation copied from SIMPL Repo and updated (if necessary)
- [x] Parameter argument variables are k_CamelCase_Key
- [x] Parameter argument strings are lower_snake_case
```
static inline constexpr StringLiteral k_AlignmentType_Key = "alignment_type";
```

## Unit Testing
- [x] 1 Unit test to instantiate the filter with the default arguments.
- [x] 1 Unit test to test output from the filter against known exemplar set of data
- [x] 1 Unit test to test invalid input code paths that are specific to a filter. Don't test that a DataPath does not exist since that test is already performed as part of the SelectDataArrayAction.

## Code Cleanup
- [x] No commented out code (rare exceptions to this is allowed..)
- [x] Filters should have both the Filter class and Algorithm class for anything beyond trivial needs
- [x] No API changes were made (or the changes have been approved)
- [x] No major design changes were made (or the changes have been approved)
- [x] Added test (or behavior not changed)
- [x] Updated API documentation (or API not changed)
- [x] Added license to new files (if any)
- [x] Added Python wrapping to new files (if any) as necessary
- [x] Added example pipelines that use the filter
- [x] Classes and methods are properly documented.


<!-- **Thanks for contributing to complex!** -->

nyoungbq committed:
Intermitten UUID Patching

nyoungbq committed:
Added code to collect and parse UUIDs during instantiation

nyoungbq committed:
Simple code fixes from code review-waiting on mapping implementation verdict (+1 squashed commits)

Squashed commits:

[27b3c8b] update test case uuids to reflect changes

JDuffeyBQ commented on src/complex/Core/Application.hpp:
```diff
@@ -118,6 +118,18 @@ class COMPLEX_EXPORT Application
+   * @brief Returns the SIMPL filter UUID(s) [v5] from the Complex filter UUID [v4]
+   * @return std::optional<std::vector<Uuid>>
+   */
+  std::optional<std::vector<Uuid>> getSimplUuid(const Uuid& complexUuid);
```
Why does this return a vector of uuids?

nyoungbq replied:
There are multiple dream3d filters that map to a single complex filter

JDuffeyBQ commented on src/Plugins/ComplexCore/src/ComplexCore/ComplexCorePlugin.hpp:
```diff
@@ -25,4 +26,24 @@ class COMPLEXCORE_EXPORT ComplexCorePlugin : public complex::AbstractPlugin
+   * their complex counterpart
+   * @return std::map<std::string, std::string>
+   */
+  std::map<std::string, std::string> getSimplToComplexMap() const override
```
The implementation of these functions should be in the cpp file.

JDuffeyBQ commented on src/Plugins/ComplexCore/src/ComplexCore/ComplexCoreLegacyUUIDMapping.hpp:
```diff
@@ -0,0 +1,125 @@
+namespace complex
+{
+  // Regex to grep UUIDs : {\"([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})\",\s\"([0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}(?:\s,\s[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})*)\"},\s\/\/\s(\w+)
+  static const std::map<std::string, std::string> k_SIMPL_to_ComplexCore
```
I would make these a `std::map<Uuid, Uuid>` instead. And maybe we should consider a bidirectional map (or approximation of one) so we don't have to duplicate the mapping.

Also I think we should use the `FilterTraits<T>::uuid` here to avoid duplicating the uuid definitions.

nyoungbq replied:
On the topic of bidirectional mapping I think that in this case I could do the same with two std::vectors that are equivalent by index. There are only 450 filter UUIDs and about 3 instances of multiple UUIDs. If it was imlemented this way we could make them std::vector<complex::Uuid> because it wouldn't require each key to be unique. Using some index tricks multiple SIMPL UUIDs could be pulled by searching for duplicate instances in the complex vector...but this introduces the chance for unwanted duplicates [especially since they are UUID version 4 (puesdo-random)].

JDuffeyBQ commented on src/complex/Core/Application.cpp:
```diff
@@ -111,6 +112,45 @@ std::filesystem::path Application::getCurrentDir() const
+std::optional<Uuid> Application::getComplexUuid(const Uuid& simplUuid)
+{
+  std::string searchUuid = simplUuid.str();
+  if(m_SIMPL_To_Complex.count(searchUuid) == 1)
```
For early returns, I think that using the error condition is better. Code with less nesting is easier to read.
```cpp
if(m_SIMPL_To_Complex.count(searchUuid) == 0)
{
  return {};
}
return Uuid::FromString(m_SIMPL_To_Complex[searchUuid]);
```
The same kind of change can be made in the other function.

JDuffeyBQ commented on src/complex/Core/Application.cpp:
```diff
@@ -171,6 +211,10 @@ void Application::loadPlugin(const std::filesystem::path& path, bool verbose)
 
   auto plugin = pluginLoader->getPlugin();
+
+  m_SIMPL_To_Complex.merge(plugin->getSimplToComplexMap());
```
`std::map::merge` skips entries that would be duplicates. We may want to report that a duplicate mapping exists.

imikejackson replied:
We should report duplicate entries of complex UUIDs on plugin load (which I think we do?) Or maybe that was in SIMPL where I hit that.

JDuffeyBQ replied:
We do report duplicate filter uuids but not in this mapping. If one plugin attempts to map a SIMPL filter to the complex one when a mapping already exists, the new mapping will be silently ignored.

imikejackson commented:
Looking at how to map back and forth I think storing the 2 lists as just 2 vectors where you know index=1 of the SIMPL vector has a direct relationship to index=1 of the COMPLEX vector. You can have duplicates in the COMPLEX vector which is an interesting case to figure out.

I think we have this mostly to go _from_ SIMPL _to_ COMPLEX. I don't see the need to go the other way so we can have duplicates in the COMPLEX vector and that is fine. the lookup time is going to be linear but these vectors are not going to be past a few hundred, maybe 500 tops in the lifetime of complex. those lookups are going to go really fast in actual wall clock time, fast enough that the performance wouldn't really matter. I'd guess on modern machines the look shouldn't take more than 10s of milliseconds so complicated multimaps and bimaps just are not needed, in my opinion. @JDuffeyBQ  thoughts?

JDuffeyBQ commented on src/Plugins/ComplexCore/src/ComplexCore/ComplexCorePlugin.hpp:
```diff
@@ -2,6 +2,7 @@
 
 #include <vector>
 
+#include "ComplexCoreLegacyUUIDMapping.hpp"
```
This can be moved to the cpp as well.

nyoungbq committed:
Rewrite Maps to support holding UUIDs and code review fixes (+4 squashed commit)

Squashed commit:

[4bc2fcf] clang formatting

[ed2df63] clang formatting

[d88b468] Rewote mapping with error checking for duplicates in applicable vector

[f779987] shuffle map include out of headers into source files

JDuffeyBQ commented on src/complex/Core/Application.cpp:
```diff
@@ -171,6 +207,30 @@ void Application::loadPlugin(const std::filesystem::path& path, bool verbose)
+      if(uuid == simplUuid)
+      {
+        throw std::runtime_error(fmt::format("Duplicate UUIDs found in the SIMPL UUID maps! UUID: {} Plugin: {}", simplUuid.str(), plugin->getName()));
+        return;
```
```suggestion
```
No need for `return` after throwing an exception.

JDuffeyBQ commented on src/complex/Core/Application.cpp:
```diff
@@ -171,6 +207,30 @@ void Application::loadPlugin(const std::filesystem::path& path, bool verbose)
+  if(m_Simpl_Uuids.size() != m_Complex_Uuids.size())
+  {
+    throw std::runtime_error(fmt::format("UUID maps are not of the same size! SIMPL UUID Vector size: {} Complex UUID Vector size: {}", m_Simpl_Uuids.size(), m_Complex_Uuids.size()));
+    return;
```
```suggestion
```

JDuffeyBQ reviewed (changes requested):
`ComplexCoreLegacyUUIDMapping.hpp` also should use `Uuid` instead of string. And potentially use the `FilterTraits<T>::uuid` for complex filters to avoid the ids from getting out of sync.

JDuffeyBQ commented on src/complex/Core/Application.hpp:
```diff
@@ -118,6 +118,18 @@ class COMPLEX_EXPORT Application
+   * @brief Returns the SIMPL filter UUID(s) [v5] from the Complex filter UUID [v4]
+   * @return std::optional<std::vector<Uuid>>
+   */
+  std::optional<std::vector<Uuid>> getSimplUuid(const Uuid& complexUuid);
```
I think it's ok to return `std::vector<Uuid>` here because there isn't a special case for an empty vector that needs to be handled. An empty vector will mean that there were no associated SIMPL filters.

nyoungbq committed:
Use filter traits in UUID maps (+1 squashed commits)

Squashed commits:

[0e5061d] Clang formatting

nyoungbq committed:
Renamed tokens according to request
